arbiter-core/src/environment.rs

```diff
@@ -101,45 +127,78 @@ impl Environment {
+    pub fn add_agent(&mut self, agent: Agent<NotAttached>) {
+        agent.attach_to_environment(self);
     }
 
     // TODO: Run should now run the agents as well as the evm.
     pub(crate) fn run(&mut self) {
         let tx_receiver = self.tx_receiver.clone();
         let mut evm = self.evm.clone();
         let event_broadcaster = self.event_broadcaster.clone();
+        let counter = Arc::clone(&self.tx_per_block);
         self.state = State::Running;
-
+        let mut expected_occurance: Option<Vec<i32>> = None;
+        if let Some(lambda) = self.lambda {
+            expected_occurance = Some(sample_poisson(lambda).unwrap());
+        }
         //give all agents their own thread and let them start watching for their evnts
         thread::spawn(move || {
             while let Ok((to_transact, tx, sender)) = tx_receiver.recv() {
                 // Execute the transaction, echo the logs to all agents, and report the execution result to the agent who made the transaction.
+                if let Some(occurance) = &expected_occurance {
+                    if counter.load(Ordering::Relaxed) >= occurance[0] as usize {
+                        evm.env.block.number += U256::from(1);
+                        counter.store(0, Ordering::Relaxed);
+                    }
+                }
+
                 evm.env.tx = tx;
                 if to_transact {
                     let execution_result = match evm.transact_commit() {
                         Ok(val) => val,
                         // URGENT: change this to a custom error
                         Err(_) => panic!("failed"),
                     };
+
                     event_broadcaster
                         .lock()
                         .unwrap()
                         .broadcast(execution_result.logs());
-                    sender.send(execution_result).unwrap();
+                    let revm_result = RevmResult {
+                        result: execution_result,
+                        block_number: convert_uint_to_u64(evm.env.block.number).unwrap(),
+                    };
+                    sender.send(revm_result).unwrap();
+                    if let Some(_occurance) = &expected_occurance {
+                        counter.fetch_add(1, Ordering::Relaxed);
+                    }
+
                 } else {
                     let execution_result = match evm.transact() {
                         Ok(val) => val,
                         // URGENT: change this to a custom error
                         Err(_) => panic!("failed"),
                     };
-                    sender.send(execution_result.result).unwrap();
+                    let result_and_block = RevmResult {
+                        result: execution_result.result,
+                        block_number: convert_uint_to_u64(evm.env.block.number).unwrap(),
+                    };
+                    sender.send(result_and_block).unwrap();
```

Comment:
Do you have a test for this new logic? I think that would be good so we know it works as intended.

Reply:
These would make the current call tests fail if they are not working